### PR TITLE
[feat]: [OIP-11]: Added connector for SignalFX

### PR DESCRIPTION
**Pull request by n00bitax**
## Checklist
- [ ] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47010)

### CODE REVIEW

**Review by naveen-harness — DISMISSED** [review]
looks good, just one nit comment

**Comment by n00bitax** [comment]
trigger utAll

File: srm-service/modules/cv-nextgen-service/contracts/delegate/src/main/java/io/harness/delegate/beans/cvng/signalfx/SignalFXUtils.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+public class SignalFXUtils {
+  public static final String HEADER_X_SF_TOKEN = "X-SF-TOKEN";
+
+  public static Map<String, String> collectionHeaders(SignalFXConnectorDTO signalFXConnectorDTO) {
+    String apiToken = new String(signalFXConnectorDTO.getApiTokenRef().getDecryptedValue());
+    Map<String, String> headers = new HashMap<>();
+    headers.put(HEADER_X_SF_TOKEN, String.format("%s", apiToken));
```
Comment (AbhijithHarness) [inline srm-service/modules/cv-nextgen-service/contracts/delegate/src/main/java/io/harness/delegate/beans/cvng/signalfx/SignalFXUtils.java:21]: nit:Why this? **String.format("%s", apiToken)**

**Review by AbhijithHarness — APPROVED** [review]

File: 440-connector-nextgen/src/main/java/io/harness/connector/mappers/signalfxmapper/SignalFXDTOToEntity.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ */
+
+package io.harness.connector.mappers.signalfxmapper;
+
+import io.harness.connector.entities.embedded.signalfxconnector.SignalFXConnector;
+import io.harness.connector.mappers.ConnectorDTOToEntityMapper;
+import io.harness.delegate.beans.connector.signalfxconnector.SignalFXConnectorDTO;
+import io.harness.encryption.SecretRefHelper;
+
+public class SignalFXDTOToEntity implements ConnectorDTOToEntityMapper<SignalFXConnectorDTO, SignalFXConnector> {
```
Comment (dhruv-harness) [inline 440-connector-nextgen/src/main/java/io/harness/connector/mappers/signalfxmapper/SignalFXDTOToEntity.java:15]: Should this not be named SignalFXDTOToEntityMapper?

File: 440-connector-nextgen/src/test/java/io/harness/connector/impl/SignalFXConnectorTest.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+@Slf4j
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+@FieldDefaults(level = AccessLevel.PRIVATE)
+@OwnedBy(HarnessTeam.CV)
+public class SignalFXConnectorTest extends CategoryTest {
+  @Mock ConnectorMapper connectorMapper;
+  @Mock ConnectorRepository connectorRepository;
+  @Mock ConnectorEntityReferenceHelper connectorEntityReferenceHelper;
+  @Mock private Map<String, ConnectionValidator> connectionValidatorMap;
```
Comment (dhruv-harness) [inline 440-connector-nextgen/src/test/java/io/harness/connector/impl/SignalFXConnectorTest.java:60]: nit: We do not need private as the annotation is already added

File: 440-connector-nextgen/src/test/java/io/harness/connector/mappers/signalfxconnectormapper/SignalFXDTOToEntityTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+import java.util.Collections;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.mockito.InjectMocks;
+import org.mockito.MockitoAnnotations;
+
+public class SignalFXDTOToEntityTest extends CategoryTest {
+  static final String urlWithoutSlash = "https://api.us1.signalfx.com";
```
Comment (dhruv-harness) [inline 440-connector-nextgen/src/test/java/io/harness/connector/mappers/signalfxconnectormapper/SignalFXDTOToEntityTest.java:30]: nit: access modifiers

File: 954-connector-beans/src/test/java/io/harness/delegate/beans/connector/UrlAnnotationTest.java [hunk 5]
```diff
@@ -223,6 +224,16 @@ public void test_PrometheusConnectorDTO_url()
     testClassField(clazz, field);
   }
 
+  @Test
+  @Owner(developers = SHREYAS)
```
Comment (dhruv-harness) [inline 954-connector-beans/src/test/java/io/harness/delegate/beans/connector/UrlAnnotationTest.java:228]: developer name?

File: 950-delegate-tasks-beans/src/main/resources/io/harness/delegate/beans/cvng/signalfx/signalFX-validation.datacollection
Comment (dhruv-harness) [inline 950-delegate-tasks-beans/src/main/resources/io/harness/delegate/beans/cvng/signalfx/signalFX-validation.datacollection:1]: Can we reuse sample data DSL for this purpose?

Comment (n00bitax) [reply 950-delegate-tasks-beans/src/main/resources/io/harness/delegate/beans/cvng/signalfx/signalFX-validation.datacollection:1]: We do reuse onboarding and data collection DSL. But I think we shouldn't mix It with the connector DSL, as it doesn't doesn't even live in the SRM module. Its only purpose is auth + check connectivity.

File: 440-connector-nextgen/src/test/java/io/harness/connector/impl/SignalFXConnectorTest.java (hunk 3 above)
Comment (n00bitax) [reply 440-connector-nextgen/src/test/java/io/harness/connector/impl/SignalFXConnectorTest.java:60]: removed

File: 440-connector-nextgen/src/test/java/io/harness/connector/mappers/signalfxconnectormapper/SignalFXDTOToEntityTest.java (hunk 4 above)
Comment (n00bitax) [reply 440-connector-nextgen/src/test/java/io/harness/connector/mappers/signalfxconnectormapper/SignalFXDTOToEntityTest.java:30]: fixed

File: 440-connector-nextgen/src/main/java/io/harness/connector/mappers/signalfxmapper/SignalFXDTOToEntity.java (hunk 2 above)
Comment (n00bitax) [reply 440-connector-nextgen/src/main/java/io/harness/connector/mappers/signalfxmapper/SignalFXDTOToEntity.java:15]: Followed convention of the project only.

File: 440-connector-nextgen/src/main/java/io/harness/connector/entities/embedded/signalfxconnector/SignalFXConnector.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.data.annotation.Persistent;
+import org.springframework.data.annotation.TypeAlias;
+
+@Value
+@Builder
+@EqualsAndHashCode(callSuper = true)
+@FieldNameConstants(innerTypeName = "SignalFXConnectorKeys")
+@StoreIn(DbAliases.NG_MANAGER)
+@Entity(value = "connectors", noClassnameStored = true)
+@Persistent
```
[the remark this answers is not in the document]
Comment (n00bitax) [reply 440-connector-nextgen/src/main/java/io/harness/connector/entities/embedded/signalfxconnector/SignalFXConnector.java:30]: I have followed the convention, guess the project uses some spring data functions, need to investigate further.

**Comment by n00bitax** [comment]
trigger utAll

**Comment by n00bitax** [comment]
trigger messagecheck

**Comment by n00bitax** [comment]
trigger messagecheck

File: srm-service/modules/cv-nextgen-service/contracts/delegate/src/main/java/io/harness/delegate/beans/cvng/signalfx/SignalFXUtils.java (hunk 1 above)
Comment (n00bitax) [reply srm-service/modules/cv-nextgen-service/contracts/delegate/src/main/java/io/harness/delegate/beans/cvng/signalfx/SignalFXUtils.java:21]: will change in the next PR

**Review by AbhijithHarness — APPROVED** [review]

**Pull request merged**